plugins/pipelines/create.js

```diff
@@ -45,41 +45,46 @@ module.exports = () => ({
             }
         },
         handler: (request, reply) => {
+            const checkoutUrl = formatCheckoutUrl(request.payload.checkoutUrl);
             const pipelineFactory = request.server.app.pipelineFactory;
             const userFactory = request.server.app.userFactory;
-            const scmUrl = formatScmUrl(request.payload.scmUrl);
             const username = request.auth.credentials.username;
 
             // fetch the user
             return userFactory.get({ username })
-                // get the user permissions for the repo
-                .then(user => user.getPermissions(scmUrl)
-                    // if the user isn't an admin, reject
-                    .then((permissions) => {
-                        if (!permissions.admin) {
-                            throw boom.unauthorized(
-                              `User ${username} is not an admin of this repo`);
-                        }
-                    })
-                    // see if there is already a pipeline
-                    .then(() => pipelineFactory.get({ scmUrl }))
-                    // if there is already a pipeline for the scmUrl, reject
-                    .then((pipeline) => {
-                        if (pipeline) {
-                            throw boom.conflict('scmUrl needs to be unique');
-                        }
-                    })
-                    // set up pipeline admins, and create a new pipeline
-                    .then(() => {
-                        const pipelineConfig = hoek.applyToDefaults(request.payload, {
-                            admins: {
-                                [username]: true
-                            },
-                            scmUrl
-                        });
-
-                        return pipelineFactory.create(pipelineConfig);
+                .then(user => user.unsealToken()
+                    .then(token => pipelineFactory.scm.parseUrl({
+                        checkoutUrl,
+                        token
                     }))
+                    // get the user permissions for the repo
+                    .then(scmUri => user.getPermissions(scmUri)
+                        // if the user isn't an admin, reject
+                        .then((permissions) => {
+                            if (!permissions.admin) {
+                                throw boom.unauthorized(
+                                  `User ${username} is not an admin of this repo`);
+                            }
+                        })
+                        // see if there is already a pipeline
+                        .then(() => pipelineFactory.get({ scmUri }))
+                        // if there is already a pipeline for the checkoutUrl, reject
+                        .then((pipeline) => {
+                            if (pipeline) {
+                                throw boom.conflict('checkoutUrl needs to be unique');
```

Comment:
Not sure if this is the correct wording since checkoutUrl is not necessarily unique. Can we say something like 'pipeline already exists'?